File: src/main/java/com/back/mozu/domain/reservation/controller/ApiV1ReservationController.java
package com.back.mozu.domain.reservation.controller;

import com.back.mozu.domain.customer.entity.Customer;
import com.back.mozu.domain.reservation.dto.ReservationDto;
import com.back.mozu.domain.reservation.service.ReservationService;
import com.back.mozu.global.response.Rq;
import com.back.mozu.global.response.RsData;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/my/reservations")
public class ApiV1ReservationController {

    private final ReservationService reservationService;
    private final Rq rq;

    @GetMapping("/")
    public RsData<List<ReservationDto.Response>> getMyReservations() {

        // Rq를 통해 현재 유저 확보
        Customer actor = rq.getActor();

        // 로그인 안 된 상태면 바로 쳐내기
        if (actor == null) {
            return RsData.of("401", "로그인이 필요한 서비스입니다.", null);
        }

        // 내 예약 목록 가져오기
        List<ReservationDto.Response> myReservations = reservationService.getMyReservation(actor.getId());

        // 유저들의 모든 예약을 리스트 형태로 RsData 담아서 전달
        return new RsData<>(
                "예약 목록 조회에 성공했습니다.",
                "200",
                myReservations
        );
    }

    @PatchMapping("/{reservationId}")
    public RsData<ReservationDto.Response> modifyMyReservation(
            @PathVariable UUID reservationId,
            @Valid @RequestBody ReservationDto.Request request) {

        // Rq를 통해 현재 유저 확보
        Customer actor = rq.getActor();

        // 보안 방어: 로그인 여부 체크
        if (actor == null) {
            return RsData.of("401", "로그인이 필요한 서비스입니다.", null);
        }

        // 수정 서비스 로직 실행 후 결과물 받아오기
        ReservationDto.Response modifiedReservation =
                reservationService.modifyMyReservation(reservationId, actor.getId(), request);

        // RsData 담아서 전달
        return new RsData<>(
                "예약 수정에 성공했습니다.",
                "200",
                modifiedReservation
        );
    }

    @PostMapping("/{reservationId}/cancel")
    public RsData<ReservationDto.Response> cancelMyReservation(@PathVariable UUID reservationId,@RequestBody ReservationDto.CancelRequest request) {

        // Rq를 통해 현재 유저 확보
        Customer actor = rq.getActor();

        // 보안 방어: 로그인 여부 체크
        if (actor == null) {
            return RsData.of("401", "로그인이 필요한 서비스입니다.", null);
        }

        // 취소 서비스 로직 실행
        ReservationDto.Response cancelledReservation =
                reservationService.cancelMyReservation(actor.getId(), reservationId, request.cancelReason());

        // RsData 담아서 전달
        return new RsData<>(
                "예약 취소에 성공했습니다.",
                "200",
                cancelledReservation
        );
    }
}

File: src/main/java/com/back/mozu/domain/reservation/controller/ApiV1ReservationController.kt
package com.back.mozu.domain.reservation.controller

import com.back.mozu.domain.reservation.dto.ReservationDto
import com.back.mozu.domain.reservation.dto.ReservationDto.CancelRequest
import com.back.mozu.domain.reservation.service.ReservationService
import com.back.mozu.global.response.Rq
import com.back.mozu.global.response.RsData
import jakarta.validation.Valid
import org.springframework.web.bind.annotation.*
import java.util.UUID

@RestController
@RequestMapping("/api/v1/my/reservations")
class ApiV1ReservationController(
    private val reservationService: ReservationService,
    private val rq: Rq,
) {

    @GetMapping("/")
    fun getMyReservations(): RsData<List<ReservationDto.Response>> {
        val actor = rq.getActor()
            ?: return RsData("로그인이 필요한 서비스입니다.", "401")
        val actorId = actor.id
            ?: return RsData("사용자 정보를 찾을 수 없습니다.", "401")
        val myReservations = reservationService.getMyReservation(actorId)
        return RsData("예약 목록 조회에 성공했습니다.", "200", myReservations)
    }

    @PatchMapping("/{reservationId}")
    fun modifyMyReservation(
        @PathVariable reservationId: UUID,
        @Valid @RequestBody request: ReservationDto.Request,
    ): RsData<ReservationDto.Response> {
        val actor = rq.getActor()
            ?: return RsData("로그인이 필요한 서비스입니다.", "401")
        val modifiedReservation = reservationService.modifyMyReservation(reservationId, actor.id, request)
        return RsData("예약 수정에 성공했습니다.", "200", modifiedReservation)
    }

    @PostMapping("/{reservationId}/cancel")
    fun cancelMyReservation(
        @PathVariable reservationId: UUID,
        @RequestBody request: CancelRequest,
    ): RsData<ReservationDto.Response> {
        val actor = rq.getActor()
            ?: return RsData("로그인이 필요한 서비스입니다.", "401")
        val cancelledReservation = reservationService.cancelMyReservation(actor.id, reservationId, request.cancelReason)
        return RsData("예약 취소에 성공했습니다.", "200", cancelledReservation)
    }
}
